### PR TITLE
Add Scarb command abstraction

### PR DESCRIPTION
**Pull request by maciektr**
Closes #

## Introduced changes
Quick idea - wrap Scarb command calls in an abstraction. 
Benefits - single points to handle things like finding Scarb binary with env var etc. 
lmkwyt 

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by MaksymilianDemitraszek** [comment]
:love

**Comment by maciektr** [comment]
Thx, added!
Do you happen to know what is wrong with this one?
```
failures:

---- e2e::declare::test_happy_case stdout ----
thread 'e2e::declare::test_happy_case' panicked at crates/cast/tests/helpers/fixtures.rs:151:40:
Could not serialize transaction output to HashMap: Error("expected value", line: 1, column: 4)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    e2e::declare::test_happy_case
```

**Comment by THenry14** [comment]
> Thx, added! Do you happen to know what is wrong with this one?
> 
> ```
> failures:
> 
> ---- e2e::declare::test_happy_case stdout ----
> thread 'e2e::declare::test_happy_case' panicked at crates/cast/tests/helpers/fixtures.rs:151:40:
> Could not serialize transaction output to HashMap: Error("expected value", line: 1, column: 4)
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> 
> 
> failures:
>     e2e::declare::test_happy_case
> ```

Hi Maciek, the issue seems to be that serde cannot serialize the output (it is needed to get tx hash) - it seems that there is some scarb output that previously wasn't:

output to be serialized on master:
```
[crates/cast/tests/e2e/declare.rs:39] &output = "{\n  \"class_hash\": \"3290316961179446805342658779538814657182136307648248088607527301219615038494\",\n  \"command\": \"declare\",\n  \"transaction_hash\": \"1123302179292931628098180772392007330386752154760729062679179466531759198564\"\n}\n"
test e2e::declare::test_happy_case ... ok
```

output to be serialized in your PR:
```
[crates/cast/tests/e2e/declare.rs:39] &output = "   Compiling lib(map) map v0.1.0 (/private/var/folders/rd/q9ymr6vd6w150wm7g3f36hf00000gn/T/.tmp8rQMrS/Scarb.toml)\nwarn: Sierra, textual Sierra and CASM lib targets have been disabled, Scarb will not produce anything\n   Compiling starknet-contract(map) map v0.1.0 (/private/var/folders/rd/q9ymr6vd6w150wm7g3f36hf00000gn/T/.tmp8rQMrS/Scarb.toml)\n    Finished release target(s) in 1 second\n{\n  \"class_hash\": \"3290316961179446805342658779538814657182136307648248088607527301219615038494\",\n  \"command\": \"declare\",\n  \"transaction_hash\": \"1123302179292931628098180772392007330386752154760729062679179466531759198564\"\n}\n"
```

**Review by Arcticae — APPROVED** [review]

File: crates/scarb-artifacts/src/command.rs
Comment (MaksymilianDemitraszek) [inline crates/scarb-artifacts/src/command.rs:1]: Probably `scarb-artifacts` is not a proper name anymore as it holds general scarb API. Could we rename it to `scarb-api` or something similar?

File: crates/sncast/src/starknet_commands/declare.rs [hunk 1]
```diff
@@ -33,41 +32,32 @@ pub struct Declare {
     pub nonce: Option<FieldElement>,
 }
 
+pub struct BuildConfig {
+    pub scarb_toml_path: Option<Utf8PathBuf>,
+    pub json: bool,
```
Comment (MaksymilianDemitraszek) [inline crates/sncast/src/starknet_commands/declare.rs:37]: Why `json` argument was added here?

**Review by MaksymilianDemitraszek — APPROVED** [review]
LGTM

File: crates/sncast/src/starknet_commands/declare.rs (hunk 1 above)
Comment (maciektr) [reply crates/sncast/src/starknet_commands/declare.rs:37]: We inherit stdio here (which have been suppressed before), so we need to set the output format to json on Scarb as well.

File: crates/sncast/tests/helpers/fixtures.rs [hunk 2]
```diff
@@ -145,17 +146,30 @@ pub fn default_cli_args() -> Vec<&'static str> {
...
+        match serde_json::de::from_slice::<T>(&line) {
+            Ok(t) => return t,
+            Err(_) => continue,
+        }
+    }
+    panic!("Failed to deserialize stdout to JSON");
+}
+
+#[derive(Deserialize)]
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (maciektr) [reply crates/sncast/tests/helpers/fixtures.rs:161]:
```
warning: fields `contract_address`, `class_hash`, and `command` are never read
   --> crates/sncast/tests/helpers/fixtures.rs:163:5
    |
161 | struct TransactionHashOutput {
    |        --------------------- fields in this struct
162 |     pub transaction_hash: String,
163 |     contract_address: Option<String>,
    |     ^^^^^^^^^^^^^^^^
164 |     class_hash: Option<String>,
    |     ^^^^^^^^^^
165 |     command: Option<String>,
    |     ^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default
```

File: crates/sncast/src/starknet_commands/declare.rs (hunk 1 above)
Comment (MaksymilianDemitraszek) [reply crates/sncast/src/starknet_commands/declare.rs:37]: My question is, why it is now configurable?

Comment (maciektr) [reply crates/sncast/src/starknet_commands/declare.rs:37]: Always has been - it's a cli argument https://github.com/foundry-rs/starknet-foundry/blob/03bbebdc6abd70db2a0ecb1acbd9b00ab7aded4a/crates/sncast/src/main.rs#L63-L65

It just has never been passed to Scarb, as Scarb output has been suppressed anyway.

File: crates/scarb-api/src/command.rs [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    Io(#[from] io::Error),
+    /// Error during execution of `scarb` command.
+    #[error("`scarb` exited with error")]
+    ScarbError,
+}
+
+/// A builder for `scarb` command invocation.
+#[derive(Clone, Debug, Default)]
+#[allow(clippy::struct_excessive_bools)]
+pub struct ScarbCommand {
```
Comment (piotmag769) [inline crates/scarb-api/src/command.rs:25]: I wonder if you don't want to have it as a part of scarb-ui or sth 🤔

File: crates/forge/src/init.rs [hunk 4]
```diff
@@ -74,34 +74,31 @@ fn extend_gitignore(path: &Path) -> Result<()> {
...
-        .output()
+        .run()
         .context("Failed to add snforge_std")?;
 
-    let scarb_version = Command::new("scarb")
+    let scarb_version = ScarbCommand::new()
         .arg("--version")
-        .stderr(Stdio::inherit())
+        .inherit_stderr()
+        .command()
```
Comment (piotmag769) [inline crates/forge/src/init.rs:101]: Why `command().output()` instead of `run()`?

File: crates/scarb-api/src/command.rs (hunk 3 above)
Comment (maciektr) [reply crates/scarb-api/src/command.rs:25]: Nope - we have `ScarbCommand` in `scarb_metadata`, but it's intended for simple case when you inherit stdio and run command only.

File: crates/forge/src/init.rs (hunk 4 above)
Comment (maciektr) [reply crates/forge/src/init.rs:101]: I don't think we want the parsing of output to be a `ScarbCommand` responsibility. We could do `output` as a shortcut which runs `command().output()`.

**Review by piotmag769 — APPROVED** [review]

**Pull request merged**